File: src/main/java/com/github/fe3dback/intellijgoarchlint/GoArchYamlCompletionContributor.java
package com.github.fe3dback.intellijgoarchlint;

import com.github.fe3dback.intellijgoarchlint.file.GoArchIcons;
import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.yaml.YAMLLanguage;
import org.jetbrains.yaml.psi.YAMLDocument;
import org.jetbrains.yaml.psi.YAMLKeyValue;

import java.util.Arrays;

public class GoArchYamlCompletionContributor extends CompletionContributor {

    private static final String keyWordDeps = "deps";
        private static final String keyWordVersion = "version";
        private static final String keyWordAllow = "allow";
        private static final String keyWordExclude = "exclude";
        private static final String keyWordExcludeFiles = "excludeFiles";
        private static final String keyWordComponents = "components";
        private static final String keyWordVendors = "vendors";
        private static final String keyWordCommonComponents = "commonComponents";
        private static final String keyWordCommonVendors = "commonVendors";

    private static String[] keyWordsTopLevel = {
            keyWordVersion,
            keyWordAllow,
            keyWordExclude,
            keyWordExcludeFiles,
            keyWordComponents,
            keyWordVendors,
            keyWordCommonComponents,
            keyWordCommonVendors,
            keyWordDeps,
    };

    public GoArchYamlCompletionContributor() {
        extend(
                CompletionType.BASIC,
                PlatformPatterns.psiElement().withLanguage(YAMLLanguage.INSTANCE),
                new Provider()
        );
    }

    private static LookupElementBuilder createKeyLookupElement(
            @NotNull final String propertyName
    ) {
        return LookupElementBuilder
                .create(propertyName)
                .withIcon(GoArchIcons.FILETYPE_ICON);
    }

    private static boolean isTopLevelMapping(final YAMLKeyValue keyValue) {
        return keyValue.getParent() instanceof YAMLDocument;
    }

    /** The main actor in generating completion suggestions. */
    private static class Provider extends CompletionProvider<CompletionParameters> {
        @Override
        protected void addCompletions(
                @NotNull final CompletionParameters completionParameters,
                final ProcessingContext processingContext,
                @NotNull final CompletionResultSet resultSet
        ) {
            final PsiElement element = completionParameters.getPosition();
            if (!GoArchPsiUtils.isGoArchFile(element)) {
                return;
            }

            final YAMLKeyValue keyValue = PsiTreeUtil.getParentOfType(element, YAMLKeyValue.class);

            // We must be at the very top level if there is no enclosing keyValue
            if (keyValue == null) {
                Arrays.stream(keyWordsTopLevel).forEachOrdered(s -> resultSet.addElement(createKeyLookupElement(s)));
                return;
            }


        }
    }
}

File: src/main/java/com/github/fe3dback/intellijgoarchlint/GoArchYamlCompletionContributor.kt
package com.github.fe3dback.intellijgoarchlint

import com.github.fe3dback.intellijgoarchlint.file.GoArchIcons
import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ProcessingContext
import com.intellij.util.containers.stream
import org.jetbrains.yaml.YAMLLanguage
import org.jetbrains.yaml.psi.YAMLKeyValue
import org.jetbrains.yaml.psi.YAMLMapping
import java.util.*

class GoArchYamlCompletionContributor() : CompletionContributor() {
    init {
        extend(
                CompletionType.BASIC,
                PlatformPatterns.psiElement().withLanguage(YAMLLanguage.INSTANCE),
                Provider()
        )
    }

    private class Provider : CompletionProvider<CompletionParameters>() {
        private val keyWordDependencies = "deps"
        private val keyWordVersion = "version"
        private val keyWordAllow = "allow"
        private val keyWordExclude = "exclude"
        private val keyWordExcludeFiles = "excludeFiles"
        private val keyWordComponents = "components"
        private val keyWordVendors = "vendors"
        private val keyWordCommonComponents = "commonComponents"
        private val keyWordCommonVendors = "commonVendors"

        private val keyWordProviderComponents = "mayDependOn"
        private val keyWordProviderVendors = "canUse"

        private val keyWordsTopLevel = arrayOf(
                keyWordVersion,
                keyWordAllow,
                keyWordExclude,
                keyWordExcludeFiles,
                keyWordComponents,
                keyWordVendors,
                keyWordCommonComponents,
                keyWordCommonVendors,
                keyWordDependencies
        )

        private val componentBasedSections = arrayOf(
                keyWordDependencies,
                keyWordCommonComponents
        )

        private val vendorBasedSections = arrayOf(
                keyWordCommonVendors
        )

        private fun applyCompletion(result: CompletionResultSet, name: String) {
            result.addElement(LookupElementBuilder
                .create(name)
                .withIcon(GoArchIcons.FILETYPE_ICON)
            )
        }

        override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
            val psiElement = parameters.position
            if (!GoArchPsiUtils.isGoArchFile(psiElement)) {
                return
            }

            // basic completion
            // -------------------
            val keyValue = PsiTreeUtil.getParentOfType(psiElement, YAMLKeyValue::class.java)
            if (keyValue == null) {
                addCompletionTopLevel(result)
                return
            }

            // semantic completion
            // -------------------
            val topLevelMapping = GoArchPsiUtils.getTopLevelMapping(keyValue) ?: return
            val currentSectionName = keyValue.keyText

            // sem, top - components
            if (componentBasedSections.contains(currentSectionName)) {
                addCompletionComponents(result, topLevelMapping)
            }

            if (vendorBasedSections.contains(currentSectionName)) {
                addCompletionVendors(result, topLevelMapping)
            }

            // sem - deps
            if (currentSectionName == keyWordProviderComponents) {
                addCompletionComponents(result, topLevelMapping)
            }

            if (currentSectionName == keyWordProviderVendors) {
                addCompletionVendors(result, topLevelMapping)
            }
        }

        private fun addCompletionTopLevel(result: CompletionResultSet) {
            Arrays.stream(keyWordsTopLevel).forEachOrdered {
                kw: String -> applyCompletion(result, kw)
            }
        }

        private fun addCompletionComponents(result: CompletionResultSet, topLevel: YAMLMapping) {
            val componentIds = GoArchPsiUtils.getComponentIds(topLevel)

            // todo: inner places: [deps.$.mayDependOn]
            componentIds.stream().forEachOrdered {
                componentId -> applyCompletion(result, componentId)
            }
        }

        private fun addCompletionVendors(result: CompletionResultSet, topLevel: YAMLMapping) {
            val vendorIds = GoArchPsiUtils.getVendorIds(topLevel)

            // todo: inner places: [deps.$.canUse]
            vendorIds.stream().forEachOrdered {
                vendorId -> applyCompletion(result, vendorId)
            }
        }
    }
}